src/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.cs

```diff
@@ -0,0 +1,370 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Sources;
+
+namespace System.Threading.Channels
+{
+    internal abstract class ResettableValueTaskSource
+    {
+        /// <summary>Sentinel object used in a field to indicate the operation has completed.</summary>
+        protected static readonly Action<object> s_completedSentinel = new Action<object>(s => Debug.Fail($"{nameof(ResettableValueTaskSource)}.{nameof(s_completedSentinel)} invoked."));
+
+        /// <summary>Throws an exception indicating that the operation's result was accessed before the operation completed.</summary>
+        protected static void ThrowIncompleteOperationException() =>
+            throw new InvalidOperationException(SR.InvalidOperation_IncompleteAsyncOperation);
+
+        /// <summary>Throws an exception indicating that multiple continuations can't be set for the same operation.</summary>
+        protected static void ThrowMultipleContinuations() =>
+            throw new InvalidOperationException(SR.InvalidOperation_MultipleContinuations);
+
+        /// <summary>Throws an exception indicating that the operation was used after it was supposed to be used.</summary>
+        protected static void ThrowIncorrectCurrentIdException() =>
+            throw new InvalidOperationException(SR.InvalidOperation_IncorrectToken);
+
+        /// <summary>Describes states the operation can be in.</summary>
+        public enum States
+        {
+            /// <summary>The operation has been assigned an owner.  No one else can use it.</summary>
+            Owned = 0,
+            /// <summary>Completion has been reserved.  Only the reserver is allowed to complete it.</summary>
+            CompletionReserved = 1,
+            /// <summary>The operation has completed and has had its result or error stored.</summary>
+            CompletionSet = 2,
+            /// <summary>The operation's result/error has been retrieved.  It's available for reuse.</summary>
+            Released = 3
+        }
+    }
+
+    internal abstract class ResettableValueTaskSource<T> : ResettableValueTaskSource, IValueTaskSource, IValueTaskSource<T>
+    {
+        private volatile int _state = (int)States.Owned;
+        private T _result;
+        private ExceptionDispatchInfo _error;
+        private Action<object> _continuation;
+        private object _continuationState;
+        private object _schedulingContext;
+        private ExecutionContext _executionContext;
+        private short _currentId;
+
+        public ValueTask CreateNonGenericValueTask() => new ValueTask(this, _currentId);
```

Comment:
Naming on these is a little weird.  It confused me when I saw it used, but it's really just a trivial thing -- get the ValueTask for this TaskSource.  The equivalent on TCS is just .Task, right?  Can we just use .ValueTask here, or .[Non]GenericValueTask?

Reply:
You mean just removing the "Create" and making it a property?  Sure.